### PR TITLE
[PGO] Add support for writing previous indexed format

### PR DESCRIPTION
**Pull request by teresajohnson**
Enable temporary support to ease use of new llvm-profdata with slightly
older indexed profiles after 16e74fd48988ac95551d0f64e1b36f78a82a89a2,
which bumped the indexed format for type profiling.

### CODE REVIEW

**Review by snehasish — APPROVED** [review]
lgtm

File: llvm/include/llvm/ProfileData/InstrProfWriter.h [hunk 1]
```diff
@@ -68,10 +68,18 @@ class InstrProfWriter {
   // Use raw pointer here for the incomplete type object.
   InstrProfRecordWriterTrait *InfoObj;
 
+  // Temporary support for writing the previous version of the format, to enable
+  // some forward compaitibility. Currently this suppresses the writing of the
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/include/llvm/ProfileData/InstrProfWriter.h:72]: done

File: llvm/include/llvm/ProfileData/InstrProfWriter.h [hunk 2]
```diff
@@ -68,10 +68,18 @@ class InstrProfWriter {
...
+  // TODO: Consider enabling this with future version changes as well, to ease
+  // deployment of newer versions of llvm-profdata.
+  bool WritePrevVersion = false;
+
 public:
   InstrProfWriter(bool Sparse = false,
                   uint64_t TemporalProfTraceReservoirSize = 0,
-                  uint64_t MaxTemporalProfTraceLength = 0);
+                  uint64_t MaxTemporalProfTraceLength = 0,
+                  bool WritePrevVersion = true);
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/include/llvm/ProfileData/InstrProfWriter.h:82]: It was meant to be false - I had it set to true for testing and nothing failed...switched it to false.

File: llvm/lib/ProfileData/InstrProfWriter.cpp [hunk 3]
```diff
@@ -433,6 +435,8 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
   IndexedInstrProf::Header Header;
   Header.Magic = IndexedInstrProf::Magic;
   Header.Version = IndexedInstrProf::ProfVersion::CurrentVersion;
+  if (WritePrevVersion)
+    Header.Version--;
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/lib/ProfileData/InstrProfWriter.cpp:439]: done. added an assert on the current version so we make sure this handling is updated if needed.

File: llvm/tools/llvm-profdata/llvm-profdata.cpp [hunk 4]
```diff
@@ -291,6 +291,15 @@ cl::opt<bool> DropProfileSymbolList(
     cl::desc("Drop the profile symbol list when merging AutoFDO profiles "
              "(only meaningful for -sample)"));
 
+// Temporary support for writing the previous version of the format, to enable
+// some forward compaitibility.
```
[the remark this answers is not in the document]
Comment (teresajohnson) [reply llvm/tools/llvm-profdata/llvm-profdata.cpp:295]: done

**Pull request merged**